src/mqt/problemsolver/equivalence_checking/equivalence_checking.py

```diff
@@ -116,56 +116,48 @@ def find_counter_examples(
-
-        operator = GroverOperator(oracle).decompose()
-        num_bits = operator.num_qubits
-        total_num_combinations = 2**num_bits
 
         qc = QuantumCircuit(num_bits)
-        qc.h(list(range(num_bits)))
+        qc.h(range(num_bits))
         qc.compose(operator.power(iterations).decompose(), inplace=True)
         qc.measure_all()
 
         qc = transpile(qc, simulator)
+        result = simulator.run(qc, shots=shots).result()
+        counts = result.get_counts()
 
-        job = simulator.run(qc, shots=shots)
-        result = job.result()
-        counts_dict = dict(result.get_counts())
-        counts_list = list(counts_dict.values())
-        counts_list.sort(reverse=True)
+        # Sort counts descending by value
+        sorted_counts = sorted(counts.items(), key=itemgetter(1), reverse=True)
+        counts_list = [count for _, count in sorted_counts]
 
-        counts_dict = dict(
-            sorted(counts_dict.items(), key=itemgetter(1))[::-1]
-        )  # Sort state dictionary with respect to values (counts)
-
-        found_counter_examples = []
         for i in range(int(total_num_combinations * 0.5)):
-            if (i + 1) == len(counts_list):
-                found_counter_examples_list = counts_list
-                found_counter_examples_dict = {
-                    list(counts_dict.keys())[t]: list(counts_dict.values())[t]
-                    for t in range(len(found_counter_examples_list))
-                }
-                found_counter_examples = list(found_counter_examples_dict.keys())
+            if (i + 1) == len(counts_list) or ((counts_list[i] - counts_list[i + 1]) > (counts_list[i] * delta)):
+                potential_counter_examples = [key for key, _ in sorted_counts[: i + 1]]
                 break
+        if potential_counter_examples:
+            break
 
-            diff = counts_list[i] - counts_list[i + 1]
-            if diff > counts_list[i] * delta:
-                found_counter_examples_list = counts_list[: i + 1]
-                found_counter_examples_dict = {
-                    list(counts_dict.keys())[t]: list(counts_dict.values())[t]
-                    for t in range(len(found_counter_examples_list))
-                }
-                found_counter_examples = list(found_counter_examples_dict.keys())
-                break
+    # Perform the check only if the potential counter examples are not empty
+    if potential_counter_examples:
+        # Check which of the two separated groups of counter examples are the real ones
+        real_counter_examples = verify_counter_examples(potential_counter_examples, miter)
 
-    if counter_examples is None:
-        return found_counter_examples
+        if counter_examples is None:
+            return real_counter_examples
+    else:
+        real_counter_examples = []
 
-    if sorted(found_counter_examples) == sorted(counter_examples):
+    if sorted(real_counter_examples or []) == sorted(counter_examples or []):
```

Comment:
I am not sure what this condition tries to test. My gut feeling is that the `or []` is not the most accessible way to do so. 

If you do something like `real_counter_examples = verify_counter_examples(potential_counter_examples, miter) if potential_counter_examples else []`, can't you just compare again the `real_counter_examples` with the `predetermined_counter_examples`?